File: 0242-valid-anagram/0242-valid-anagram.py

#create a frequency table for each string
#check if the frequency tables are the same by value, can do python using ==, in javascript it would require manual iteration over the strings



class Solution:
    def map_s(self, s:str)->dict:
        out = {}
        for char in s:
            out[char] = out.get(char, 0) + 1
        
        return out

    def isAnagram(self, s: str, t: str) -> bool:
        s_table = self.map_s(s)
        t_table = self.map_s(t)

        return s_table == t_table
        
        
        

File: 0242-valid-anagram/0242-valid-anagram.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */

/**
input: two strings, S and T
output: boolean, true if valid anagram, false otherwise

no edgecases here
use a gaurd, check if length is not equal -> return false

create frequency tables of both s and t

for each key: value pair in s, check if that key value pair is the
same in t
if not, return false

if the above operation is able to complete witbout a return, return true
*/

function makeFreqTable(str) {
    let out = {}

    for (const char of str) {
        out[char] = (out[char] || 0) + 1;
    }

    return out
}

const isAnagram = function(s, t) {
    if (s.length != t.length) return false
    
    const sMap = makeFreqTable(s)
    const tMap = makeFreqTable(t)

    for (const key in sMap) {
        if (sMap[key] != tMap[key]) return false
    }

    return true

    
};